### PR TITLE
[SPARK-17636][SQL][WIP] Parquet predicate pushdown in nested fields

### PR DESCRIPTION
**Pull request by dbtsai**
## What changes were proposed in this pull request?

Support Parquet predicate pushdown in nested fields

## How was this patch tested?

Existing tests and new tests are added.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala [hunk 1]
```diff
@@ -437,53 +436,63 @@ object DataSourceStrategy {
...
-
-      case expressions.LessThanOrEqual(a: Attribute, Literal(v, t)) =>
-        Some(sources.LessThanOrEqual(a.name, convertToScala(v, t)))
-      case expressions.LessThanOrEqual(Literal(v, t), a: Attribute) =>
-        Some(sources.GreaterThanOrEqual(a.name, convertToScala(v, t)))
-
-      case expressions.InSet(a: Attribute, set) =>
-        val toScala = CatalystTypeConverters.createToScalaConverter(a.dataType)
-        Some(sources.In(a.name, set.toArray.map(toScala)))
+      case expressions.EqualTo(e: Expression, Literal(v, t)) =>
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala]: This PR will be a good performance improvement for Spark 2.5.0.

**Comment by viirya** [comment]
retest this please.

**Comment by dongjoon-hyun** [comment]
Hi, @dbtsai . Could you fix the UT failure?
```scala
[info] ParquetFilterSuite:
...
[info] - SPARK-12218 Converting conjunctions into Parquet filter predicates *** FAILED *** (19 milliseconds)
[info]   Expected None, but got Some(lt(a, 10)) (ParquetFilterSuite.scala:802)
```

**Comment by dbtsai** [comment]
I'm breaking this PRs into three smaller PR. I'll fix the tests in those smaller PRs. Thanks.

**Comment by aniket486** [comment]
@dbtsai are you still working on this?

**Comment by dbtsai** [comment]
Going to close this in favor of new PR, https://github.com/apache/spark/pull/27728

**Pull request closed without merging**